Framework/Algorithms/src/MaskNonOverlappingBins.cpp

```diff
@@ -0,0 +1,136 @@
+#include "MantidKernel/MultiThreaded.h"
+
+namespace {
+/// Constants for the algorithm's property names.
+namespace Prop {
+std::string const COMPARISON_WS{"ComparisonWorkspace"};
+std::string const INPUT_WS{"InputWorkspace"};
+std::string const MASK_PARTIAL{"MaskPartiallyOverlapping"};
+std::string const OUTPUT_WS{"OutputWorkspace"};
+} // namespace Prop
+} // namespace
+
+namespace Mantid {
+namespace Algorithms {
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+DECLARE_ALGORITHM(MaskNonOverlappingBins)
+
+/// Algorithms name for identification. @see Algorithm::name
+std::string const MaskNonOverlappingBins::name() const {
+  return "MaskNonOverlappingBins";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int MaskNonOverlappingBins::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+std::string const MaskNonOverlappingBins::category() const {
+  return "Transforms\\Masking";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+std::string const MaskNonOverlappingBins::summary() const {
+  return "Marks bins in InputWorkspace which are out of the X range of the "
+         "second workspace.";
+}
+
+/// Return a list of the names of related algorithms.
+std::vector<std::string> const MaskNonOverlappingBins::seeAlso() const {
+  return {"MaskBins", "MaskBinsIf"};
+}
+
+/** Initialize the algorithm's properties.
+ */
+void MaskNonOverlappingBins::init() {
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::INPUT_WS, "", Kernel::Direction::Input),
+                  "A workspace to mask.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::OUTPUT_WS, "", Kernel::Direction::Output),
+                  "The masked workspace.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::COMPARISON_WS, "", Kernel::Direction::Input),
+                  "A workspace to compare the InputWorkspace's binning to.");
+  declareProperty(Prop::MASK_PARTIAL, false,
+                  "If true, mask also bins that overlap only partially.");
+}
+
+/// Returns a map from property name to message in case of invalid values.
+std::map<std::string, std::string> MaskNonOverlappingBins::validateInputs() {
+  std::map<std::string, std::string> issues;
+  API::MatrixWorkspace_const_sptr inputWS = getProperty(Prop::INPUT_WS);
+  API::MatrixWorkspace_const_sptr comparisonWS =
+      getProperty(Prop::COMPARISON_WS);
+  if (inputWS->getNumberHistograms() != comparisonWS->getNumberHistograms()) {
+    issues[Prop::COMPARISON_WS] =
+        "The number of histogams mismatches with " + Prop::INPUT_WS;
```

Comment:
Would there be any advantage to allowing the comparison workspace to be a single spectra? i.e. broadcasting rules of exactly same number of spectra or 1.

Reply:
For my use case, no. For general use, I don't know. I'd leave it like this until someone needs the feature to avoid over-engineering.